File: frontend/src/utils/money.test.ts
import { describe, expect, it } from 'vitest'
import { basisPointsToInput, formatCents, formatPercentage, parsePercentage, split } from './money'

describe('parsePercentage', () => {
  it.each([
    ['33.33', 3333],
    ['33,5', 3350],
    ['100', 10000],
    ['0.01', 1],
    ['  50  ', 5000],
  ])('parses %s as %i basis points', (input, want) => {
    expect(parsePercentage(input)).toBe(want)
  })

  it.each([['33.333'], ['abc'], ['-5'], ['']])('rejects %s', (input) => {
    expect(parsePercentage(input)).toBeNull()
  })
})

describe('formatCents', () => {
  it('formats cents as euros', () => {
    expect(formatCents(12050)).toBe('€120.50')
    expect(formatCents(0)).toBe('€0.00')
  })
})

describe('formatPercentage', () => {
  it('renders basis points with a % suffix', () => {
    expect(formatPercentage(3333)).toBe('33.33 %')
    expect(formatPercentage(10000)).toBe('100.00 %')
  })
})

describe('basisPointsToInput', () => {
  it('drops trailing zeros for whole percentages', () => {
    expect(basisPointsToInput(5000)).toBe('50')
  })

  it('keeps the fraction for non-whole percentages', () => {
    expect(basisPointsToInput(3333)).toBe('33.33')
  })
})

describe('split', () => {
  it('gives the leftover cent to the largest remainder', () => {
    expect(split(1000, [3333, 3333, 3334])).toEqual([333, 333, 334])
  })

  it('gives everything to a single share at 100%', () => {
    expect(split(12050, [10000])).toEqual([12050])
  })

  it('splits a zero total into zeros', () => {
    expect(split(0, [3333, 3333, 3334])).toEqual([0, 0, 0])
  })

  it.each([
    [1000, [3333, 3333, 3334]],
    [12050, [10000]],
    [0, [3333, 3333, 3334]],
    [999, [2500, 2500, 2500, 2500]],
  ])('sums to the total for total=%i basisPoints=%j', (totalCents, basisPoints) => {
    const amounts = split(totalCents, basisPoints)
    expect(amounts.reduce((sum, amount) => sum + amount, 0)).toBe(totalCents)
  })
})

File: backend/internal/money/money_test.go
package money

import (
	"testing"
)

func TestSplit(t *testing.T) {
	tests := []struct {
		name        string
		totalCents  int64
		basisPoints []int32
		want        []int64
	}{
		{
			name:        "leftover cent goes to the largest remainder",
			totalCents:  1000,
			basisPoints: []int32{3333, 3333, 3334},
			want:        []int64{333, 333, 334},
		},
		{
			name:        "single share at 100%",
			totalCents:  12050,
			basisPoints: []int32{10000},
			want:        []int64{12050},
		},
		{
			name:        "zero total",
			totalCents:  0,
			basisPoints: []int32{3333, 3333, 3334},
			want:        []int64{0, 0, 0},
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := Split(tt.totalCents, tt.basisPoints)

			if len(got) != len(tt.want) {
				t.Fatalf("Split() returned %d amounts, want %d", len(got), len(tt.want))
			}
			var sum int64
			for i, amount := range got {
				if amount != tt.want[i] {
					t.Errorf("Split()[%d] = %d, want %d", i, amount, tt.want[i])
				}
				sum += amount
			}
			if sum != tt.totalCents {
				t.Errorf("Split() amounts sum to %d, want %d", sum, tt.totalCents)
			}
		})
	}
}
